### ISSUE TITLE
Refresh account on Launch

### ISSUE THREAD

**Issue by AbyssalMonkey** [PrismLauncher/PrismLauncher#520]
### Role

I play too much modded.

### Suggestion

I want the launcher to automatically refresh my account when I launch the instance.

### Benefit

I never see the "Invalid Session Token" message again.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

**Comment by Apocalypsing** [comment]
I find I'm running into this problem again too. This issue tends to happen if Prism is left open for a considerable amount of time, as that only seems to exacerbate the issue.

**Comment by jaknz** [comment]
Plus one for me too, I tend to leave Prism open, but have to manually go in and refresh my account every couple of days. This would be an epic improvement.

**Comment by nightpool** [comment]
This would be a huge QoL improvement—it's very silly that the launcher only refreshes the session once when you first start it up

**Comment by wlhlm** [comment]
~~I'm hitting this regularly. Basically daily (effectively every play session) I have to manually refresh the account tokens in the accounts settings page. Would love to see this getting improved.~~

Turns out what I'm actually hitting is #592.

**Comment by wolsoot** [comment]
Still encountering this on Prism Launcher 8.4, not sure why it was closed.

<br>
Just to make sure I'm not barking up the wrong tree here:  

The problem is that sometimes (usually after having the launcher open for a while - think, multiple days) when I launch an instance, it starts without problems or warning, but as soon as I attempt to connect to a server, I'm greeted by the infamous "Failed to connect to server; Failed to log in: Invalid session (Try restarting your game and the launcher)".  
Simply closing the game and launching again leads to the same result.  
However going to the Accounts section in the settings, selecting the account and hitting refresh resolves the error. Launching an instance again after having completed this procedure allows me to join servers without any problem.

I was under the impression that the present issue is a feature request to automatically do this (currently manual) refreshing step when I launch an instance.

**Comment by Trial97** [comment]
It does: it is done right before the start of the game but with the condition that the account is offline and you request to launch online. If the account is not marked offline then we do nothing.
Do you perhaps have multiple online accounts?

**Comment by wolsoot** [comment]
I'm not sure what exactly it means for an account to be offline, but I suspect it's not the same thing that causes the "Invalid session" error when connecting to servers. Prism isn't showing any indication that the account is offline (launch button reads "Launch", _not_ "Launch Offline" and there's no warning in the main UI. So my guess is that this requirement on the account being "offline" before initiating a refresh is what prevents this feature from working as I would expect it to.   
There's only one account currently present in the accounts section.  

For what it's worth, the minecraft log does mention a 401 error during authentication:
```
[22:31:59] [Render thread/INFO]: Environment: authHost='https://authserver.mojang.com', accountsHost='https://api.mojang.com', sessionHost='https://sessionserver.mojang.com', servicesHost='https://api.minecraftservices.com', name='PROD'
[22:31:59] [Render thread/INFO]: Mixing client.MixinYggdrasilUserApiService from mixins/common/nochatreports.mixins.json into com.mojang.authlib.yggdrasil.YggdrasilUserApiService
[22:32:00] [Render thread/ERROR]: Failed to verify authentication
com.mojang.authlib.exceptions.InvalidCredentialsException: Status: 401
	at com.mojang.authlib.exceptions.MinecraftClientHttpException.toAuthenticationException(MinecraftClientHttpException.java:56) ~[authlib-4.0.43.jar:?]
	at com.mojang.authlib.yggdrasil.YggdrasilUserApiService.fetchProperties(YggdrasilUserApiService.java:156) ~[authlib-4.0.43.jar:?]
	at com.mojang.authlib.yggdrasil.YggdrasilUserApiService.<init>(YggdrasilUserApiService.java:55) ~[authlib-4.0.43.jar:?]
	at com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService.createUserApiService(YggdrasilAuthenticationService.java:124) ~[authlib-4.0.43.jar:?]
	at net.minecraft.class_310.method_31382(class_310.java:733) ~[client-intermediary.jar:?]
	at net.minecraft.class_310.<init>(class_310.java:442) ~[client-intermediary.jar:?]
	at net.minecraft.client.main.Main.main(Main.java:211) ~[minecraft-1.20-client.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:468) ~[fabric-loader-0.14.22.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74) ~[fabric-loader-0.14.22.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23) ~[fabric-loader-0.14.22.jar:?]
	at org.prismlauncher.launcher.impl.StandardLauncher.launch(StandardLauncher.java:100) ~[NewLaunch.jar:?]
	at org.prismlauncher.EntryPoint.listen(EntryPoint.java:129) ~[NewLaunch.jar:?]
	at org.prismlauncher.EntryPoint.main(EntryPoint.java:70) ~[NewLaunch.jar:?]
Caused by: com.mojang.authlib.exceptions.MinecraftClientHttpException: Status: 401
	at com.mojang.authlib.minecraft.client.MinecraftClient.readInputStream(MinecraftClient.java:85) ~[authlib-4.0.43.jar:?]
	at com.mojang.authlib.minecraft.client.MinecraftClient.get(MinecraftClient.java:48) ~[authlib-4.0.43.jar:?]
	at com.mojang.authlib.yggdrasil.YggdrasilUserApiService.fetchProperties(YggdrasilUserApiService.java:129) ~[authlib-4.0.43.jar:?]
	... 11 more
```  

I doubt it has anything to do with my mods, but I can try to reproduce with a vanilla instance if need be.  

I'm not sure why the decision was made to not always refresh on launch, but if there's some reason to prefer that as the default behavior, it would be nice if we could at least explicitly configure Prism to always refresh, no matter if the account reads as offline or not.

**Comment by Trial97** [comment]
@wolsoot can you test a build for me: 
- PR: https://github.com/PrismLauncher/PrismLauncher/pull/2571
- assets: https://github.com/PrismLauncher/PrismLauncher/actions/runs/9674913172?pr=2571

I added an extra check if the refresh should happen before the launch of the game.

**Comment by wolsoot** [comment]
Will do, however I can't promise timely feedback since it usually takes a while for the problem to manifest (I don't know how long exactly), unless there's some easy manual way to invalidate the session. For the same reason it would be hard for me to have conclusive evidence that it's working.  

I guess one way would be to also keep the instance alive indefinitely and wait until connecting to a server gives the invalid session error, then close the instance and launch it again. The acceptance criteria would be that after relaunching, the error goes away and I can connect to a server, right?

**Comment by Trial97** [comment]
> I guess one way would be to also keep the instance alive indefinitely and wait until connecting to a server gives the invalid session error, then close the instance and launch it again. The acceptance criteria would be that after relaunching, the error goes away and I can connect to a server, right?

Yes,  the token expires after 24 hours and you need to restart the game.

**Comment by wolsoot** [comment]
Unfortunately no cigar.  
I got the session error while trying to connect to a server, so I relaunched the instance, but it didn't fix the issue.  

Maybe there's some problem with how `shouldRefresh()` is calculated? I had a look at my accounts.json, and the yggdrasil token expiration date is 1719473176 (about an hour ago).

**Comment by Trial97** [comment]
It's strange that with that specific build, I get an account refresh on each launch(due to another bug the validity is always assumed; this I gonna fix elsewhere).
So I don't understand how in your case the refresh is not called on instance launch.
Do you still have the prism logs from that session?

**Comment by wolsoot** [comment]
Maybe the validity is None instead of Assumed when the token is expired? I'm guessing you didn't leave prism open for 24h to let the token expire.

I built prism from source and added some logging statements, but I'll have to wait until tomorrow/Saturday for the session to expire.

I still haven't closed the prism launcher (the one from the build you linked) since I started it yesterday, so it's all in one big log file, I'll attach it here.
[PrismLauncher-0_shouldRefresh.log](https://github.com/user-attachments/files/16015654/PrismLauncher-0_shouldRefresh.log)

**Comment by Trial97** [comment]
If the validity is None then that means the refresh will not work and you would need to manually login in again(as the refresh token would be empty). If you encounter problems like the account being in a state of Working forever we may need to rebase over another PR: https://github.com/PrismLauncher/PrismLauncher/pull/2504 
And if you do add logs go nuts with them: add one at each return statement of `shouldRefresh` function and make them visible.

**Comment by wolsoot** [comment]
Unless I'm getting my wires crossed somewhere, None validity is exactly what's happening in my case.
Or rather, getting validity=None is possible while still having a logged-in account that is able to connect to servers.
Have a look at my log file. Custom log statements are prepended with `////>>`.  
Since my session is not yet expired, I'm able to log into servers in this current state.  

I'm guessing I was able to produce this state by deleting `accounts.json` before starting prism (notice the "Failed to read the account list file (accounts.json)." in the log file)~~, but it should be possible to get into a similar state without doing so, since I encountered what I assume is similar behavior on the regular prism launcher without touching the file (though none of my "real" prism logs contain the mentioned critical message, so it must manifest through some other mechanism)~~.

P.S. I did encounter the endless Working, and tried to login using the "Legacy" option for a while, but in my latest test MS did accept my "Recommended" login for one reason or another.
[PrismLauncher-0_extra_logging.log](https://github.com/user-attachments/files/16017633/PrismLauncher-0_extra_logging.log)

**Comment by wolsoot** [comment]
I'm afraid I likely had deleted the accounts.json before testing your build as well, which is likely what caused the behavior I reported 7 hours ago. So please disregard that result for now.  

I've now restarted both my logged-up build of prism and the one from your PR and will check back in tomorrow, but my assumption right now is that #2571 fixes this issue (your observation that it over-corrects somewhat notwithstanding).

**Comment by Trial97** [comment]
Did you have any luck with this or should we postpone it until the login bugs are fixed?

**Comment by wolsoot** [comment]
My token is set to expire in about 4 hours, so that's the earliest I'll have any feedback. Might be a few hours more depending on my availability.
I expect a positive result since you already tested it, the commit looks reasonable, and I have a solid theory of why it wasn't working in my previous test and have good reason to believe I've avoided that mistake this time around.

**Comment by wolsoot** [comment]
It's working for me.
Session expired, relaunched the game, logged in to server without error.

Also noticed the behavior of it always refreshing on launch, even if the token isn't expired (or close to expiring) yet.

**Comment by Trial97** [comment]
Good to know that that works I'm gonna merge that PR then.

> Also noticed the behavior of it always refreshing on launch, even if the token isn't expired (or close to expiring) yet.

That is fixed by https://github.com/PrismLauncher/PrismLauncher/pull/2504
